Revisiting mkimage-arch.sh

ztombol opened the pull request:
# Overview

This pull request aims to clean up and slightly reduce the size of Arch Linux base images generated with `mkimage-arch.sh`. So far it does the following two things:
- prevent installing more unnecessary packages
- delete man pages of installed packages

These changes reduce the image size by 17.7 MB or ~5% (from 310.4 MB to 292.7 MB).

**_Note:**_ _Image size is the VIRTUAL SIZE as reported by `docker inspect` on a btrfs file system. The test images were built around 2015-01-08 18:45:48 UTC._
# Details
## Deleting man pages

The script already skips installing [`man-db`](https://www.archlinux.org/packages/core/x86_64/man-db/) and [`man-pages`](https://www.archlinux.org/packages/core/x86_64/man-pages/) to save on disk space, however it fails to delete man pages that belong to the installed packages. `rm -r /usr/share/man/*` accomplishes just that, and reduces the image size by about 11.9 MB.
## Revising list of installed packages

This is more for cleaning up than to save a few megabytes, but it also further reduces the image size by 5.8 MB.

The following is a revised list of packages that should not be installed and a short explanation on why they should be left out, along with a few notes on packages that cannot be left out. Packages denoted with an _*_ are already ignored in the current version of `mkimage-arch.sh`, the rest are new additions.
- [`linux*`](https://www.archlinux.org/packages/core/x86_64/linux/), [`systemd-sysvcompat`](https://www.archlinux.org/packages/core/x86_64/systemd-sysvcompat/) - Containers do not boot.
  
  > [`systemd`](https://www.archlinux.org/packages/core/x86_64/systemd/) - [`procps-ng`](https://www.archlinux.org/packages/core/x86_64/procps-ng/) (`pkill`, `pgrep`, etc.) depends on it.
- [`openresolv`](https://www.archlinux.org/packages/core/x86_64/openresolv/) - `resolv.conf` is managed by the docker daemon.
- [`netctl`](https://www.archlinux.org/packages/core/x86_64/netctl/),[`dhcpcd`](https://www.archlinux.org/packages/core/x86_64/dhcpcd/),[`iproute2`](https://www.archlinux.org/packages/core/x86_64/iproute2/) - Network interface is managed by the docker daemon.
  
  > [`inetutils`](https://www.archlinux.org/packages/core/x86_64/inetutils/) - Contains `telnet` and `ftp`.
  > [`iputils`](https://www.archlinux.org/packages/core/x86_64/iputils/) - Contains `ping`.
- [`pciutils*`](https://www.archlinux.org/packages/core/x86_64/pciutils/), [`pcmciautils*`](https://www.archlinux.org/packages/core/x86_64/pcmciautils/), [`usbutils`](https://www.archlinux.org/packages/core/x86_64/usbutils/) - Hardware management is done by the host.
- [`jfsutils*`](https://www.archlinux.org/packages/core/x86_64/jfsutils/), [`xfsprogs*`](https://www.archlinux.org/packages/core/x86_64/xfsprogs/), [`reiserfsprogs*`](https://www.archlinux.org/packages/core/x86_64/reiserfsprogs/) - File systems are managed by the host.
  
  > [`e2fsprogs`](https://www.archlinux.org/packages/core/x86_64/e2fspriogs/) - [`pam`](https://www.archlinux.org/packages/core/x86_64/pam/), a `base` package indirectly depends on it.
- [`lvm2*`](https://www.archlinux.org/packages/core/x86_64/lvm2/), [`mdadm*`](https://www.archlinux.org/packages/core/x86_64/mdadm/), [`cryptsetup*`](https://www.archlinux.org/packages/core/x86_64/cryptsetup/), [`device-mapper`](https://www.archlinux.org/packages/core/x86_64/device-mapper/) - Storage devices are managed by the host.
- [`man-db*`](https://www.archlinux.org/packages/core/x86_64/man-db/), [`man-pages*`](https://www.archlinux.org/packages/core/x86_64/man-pages/) - Documentation not useful inside a container.
  
  > [`texinfo`](https://www.archlinux.org/packages/core/x86_64/texinfo/) - `pacman` indirectly depends on it.
  > [`licenses`](https://www.archlinux.org/packages/core/x86_64/licenses/) - Should not be left out, see [here](https://wiki.archlinux.org/index.php/PKGBUILD#license)
- [`s-nail*`](https://www.archlinux.org/packages/core/x86_64/s-nail/), [`nano`](https://www.archlinux.org/packages/core/x86_64/nano/), [`vi`](https://www.archlinux.org/packages/core/x86_64/vi/) - Unlikely to be widely used.

For completeness, here are the members of the `base` group that are installed.
[`bash`](https://www.archlinux.org/packages/core/x86_64/bash/), [`bzip2`](https://www.archlinux.org/packages/core/x86_64/bzip2/), [`coreutils`](https://www.archlinux.org/packages/core/x86_64/coreutils/), [`diffutils`](https://www.archlinux.org/packages/core/x86_64/diffutils/), [`e2fsprogs`](https://www.archlinux.org/packages/core/x86_64/e2fspriogs/), [`file`](https://www.archlinux.org/packages/core/x86_64/file/), [`filesystem`](https://www.archlinux.org/packages/core/x86_64/filesystem/), [`findutils`](https://www.archlinux.org/packages/core/x86_64/findutils/), [`gawk`](https://www.archlinux.org/packages/core/x86_64/gawk/), [`gcc-libs`](https://www.archlinux.org/packages/core/x86_64/gcc-libs/), [`gettext`](https://www.archlinux.org/packages/core/x86_64/gettext/), [`glibc`](https://www.archlinux.org/packages/core/x86_64/glibc/), [`grep`](https://www.archlinux.org/packages/core/x86_64/grep/), [`gzip`](https://www.archlinux.org/packages/core/x86_64/gzip/), [`inetutils`](https://www.archlinux.org/packages/core/x86_64/inetutils/), [`iputils`](https://www.archlinux.org/packages/core/x86_64/iputils/), [`less`](https://www.archlinux.org/packages/core/x86_64/less/), [`licenses`](https://www.archlinux.org/packages/core/x86_64/licenses/), [`logrotate`](https://www.archlinux.org/packages/core/x86_64/logrotate/), [`pacman`](https://www.archlinux.org/packages/core/x86_64/pacman/), [`perl`](https://www.archlinux.org/packages/core/x86_64/perl/), [`procps-ng`](https://www.archlinux.org/packages/core/x86_64/procps-ng/), [`psmisc`](https://www.archlinux.org/packages/core/x86_64/psmisc/), [`sed`](https://www.archlinux.org/packages/core/x86_64/sed/), [`shadow`](https://www.archlinux.org/packages/core/x86_64/shadow/), [`sysfsutils`](https://www.archlinux.org/packages/core/x86_64/sysfsutils/), [`tar`](https://www.archlinux.org/packages/core/x86_64/tar/), [`texinfo`](https://www.archlinux.org/packages/core/x86_64/texinfo/), [`util-linux`](https://www.archlinux.org/packages/core/x86_64/util-linux/), [`which`](https://www.archlinux.org/packages/core/x86_64/which/).

**_Note:**_ _`groff` has been removed from the ignore list because it is actually not installed by `base` (it's in `base-devel`)._
## Images sizes

The following table summarises the space saving introduced in this pull request. Images are _cumulative_ in that an image also contains the changes made by the one above it, e.g. _minimal packages_ also deletes the man pages. Differences in size are calculated against the image immediately above.

| image | size (MB) | diff (MB) |
| :-- | --: | --: |
| original | 310.4 |  |
| deleting man | 298.5 | -11.9 |
| minimal packages | 294.3 | -4.2 |
| removed editors | 292.7 | -1.6 |
# Feedback wanted

The patched script builds a working image, however it's possible that I overlooked an unnecessary package or removed an important one. I linked all package names to the respective pages on the Arch Linux web site to make it easier to review the changes. Let me know what you think.

<!-- References -->

<!-- IGNORED -->

<!-- INSTALLED -->

ztombol committed:
Revise list of installed packages in mkimage-arch.sh

Docker-DCO-1.1-Signed-off-by: Zoltan Tombol <zoltan.tombol@gmail.com> (github: ztombol)

ztombol committed:
Delete man pages of installed packages in mkimage-arch.sh

Docker-DCO-1.1-Signed-off-by: Zoltan Tombol <zoltan.tombol@gmail.com> (github: ztombol)

jfrazelle commented:
I personally think the images should be clean as well and not include editors etc, but @tianon is the main deciding factor here

tianon commented:
I'm definitely +1 on removing editors, etc. from base images, but sometimes upstream isn't as privy to that (see Ubuntu as an example).

SvenDowideit commented:
awesome info - bookmarking for docs :)

ztombol commented:
Thanks guys! I updated the pull request to remove `nano` and `vi` (shaving off another 1.6 MB) and updated the OP to reflect the changes.

@tianon let me know if you want me to squash the commits into one before merging.

tianon commented on contrib/mkimage-arch.sh:
```diff
@@ -18,7 +18,12 @@ ROOTFS=$(mktemp -d ${TMPDIR:-/var/tmp}/rootfs-archlinux-XXXXXXXXXX)
+pcmciautils,usbutils,jfsutils,xfsprogs,reiserfsprogs,lvm2,mdadm,cryptsetup,\
+device-mapper,man-db,man-pages,s-nail,nano,vi
+EOF
+)
```
Wouldn't this be easier to maintain long-term if we used something more like:
```bash
PKGIGNORE=(
	cryptsetup
	device-mapper
	dhcpcd
	iproute2
	jfsutils
	linux
	lvm2
	man-db
	man-pages
	mdadm
	nano
	netctl
	openresolv
	pciutils
	pcmciautils
	reiserfsprogs
	s-nail
	systemd-sysvcompat
	usbutils
	vi
	xfsprogs
)
IFS=','
PKGIGNORE="${PKGIGNORE[*]}"
unset IFS
```

dasJ commented:
With the list of @tianon, those packages are being installed into the image:
```bash
bzip2
coreutils
diffutils
e2fsprogs
file
filesystem
findutils
gawk
gcc-libs
gettext
glibc
grep
gzip
inetutils
iputils
less
licenses
logrotate
pacman
perl
procps-ng
psmisc
sed
shadow
sysfsutils
tar
texinfo
util-linux
which
```

If we ignore the fact, that dockerfiles may depend on them, or they may be useful for most of the users, these packages could also be removed: `diffutils`, `file`, `gettext`, `grep`, `inetutils`, `iputils`, `logrotate`, `psmisc`, `sysfsutils`, `tar`, and `which`.

`procps-ng` has to be removed after the pacman keys were populated because `procps-ng` contains `pkill`. Removing this package also removes the `systemd` package.

`locale-gen`, which is called from the script requires `sed`.

When leaving `procps-ng` and `sed` in, my image is 285,4MB in size.

ztombol commented:
@tianon good point. Amended.

tianon commented:
@dasJ so, in both Debian and Gentoo, there's an understanding in package metadata that packages in the base set are not required to be listed explicitly in dependencies of other packages -- is that true in Arch as well?  ie, are packages allowed to assume that basic tools like `pkill`, `sed`, `grep`, etc are available without explicitly depending on them?

tianon commented:
This PR as-is LGTM anyhow, but I think there's clearly room for further discussion about the exist list of packages to purge. :+1:

ztombol commented:
@dasJ You left out `bash` from the list, but other than that it's correct.

> "If we ignore the fact, that dockerfiles may depend on them, or they may be useful for most of the users..."

I think this is the perfect argument against removing those packages.

`file`, `inetutils`, `iputils`, `procps-ng`, `ps-misc`, `sed`, `sysfsutils`, `tar`, `which` provide binaries that are too essential to be removed. An image without them would barely work. It would be impossible to derive new images from such a base. Many shell scripts would fail as well.

I'm not sure about `logrotate`, `diffutils` and `gettext`.

PS.: The list of installed packages in the OP was not correct. I missed `e2fsprogs` and `texinfo`, and included `iproute2`. It should be fixed now.

jfrazelle commented:
LGTM

ztombol commented:
@tianon That is not true for Arch.

Packages list all run-time dependencies even if they are in `base`. For example [`device-mapper`][pkg-device-mapper] lists `systemd`, both of which are in `base`. And so does the [`docker`][pkg-docker] package. I think that's because in Arch `base` is a suggestion rather than a requirement and users are encouraged to customise even by leaving out packages they don't need (e.g. `jfsutils`).

*Note: There is a similar rule for packages built from source. They [should not list compile time dependencies][aw-makedepends] that are already in the `base-devel` group (packages required for building packages, e.g. `make` and `gcc`).*

[pkg-device-mapper]: https://www.archlinux.org/packages/core/x86_64/device-mapper/
[pkg-docker]: https://www.archlinux.org/packages/community/x86_64/docker/
[aw-makedepends]: https://wiki.archlinux.org/index.php/PKGBUILD#makedepends

tianon commented:
Ah cool, thanks for clarifying!
